xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java

```diff
@@ -142,32 +162,51 @@ private static void validateWatcherAndPush(
      * This method must be executed in a blocking thread because
      * {@link KubernetesEndpointGroupBuilder#build()} blocks the execution thread.
      */
-    public static KubernetesEndpointGroup createKubernetesEndpointGroup(ServiceEndpointWatcher watcher) {
+    public static CompletableFuture<KubernetesEndpointGroup> createKubernetesEndpointGroup(
+            ServiceEndpointWatcher watcher, MetaRepository metaRepository, Executor executor) {
         final KubernetesConfig kubernetesConfig = watcher.getKubernetesConfig();
         final String serviceName = watcher.getServiceName();
 
-        final KubernetesEndpointGroupBuilder kubernetesEndpointGroupBuilder =
-                KubernetesEndpointGroup.builder(toConfig(kubernetesConfig)).serviceName(serviceName);
-        if (!isNullOrEmpty(kubernetesConfig.getNamespace())) {
-            kubernetesEndpointGroupBuilder.namespace(kubernetesConfig.getNamespace());
-        }
-        if (!isNullOrEmpty(watcher.getPortName())) {
-            kubernetesEndpointGroupBuilder.portName(watcher.getPortName());
-        }
-
-        return kubernetesEndpointGroupBuilder.build();
+        return toConfig(kubernetesConfig, metaRepository).thenApplyAsync(config -> {
+            final KubernetesEndpointGroupBuilder kubernetesEndpointGroupBuilder =
+                    KubernetesEndpointGroup.builder(config).serviceName(serviceName);
+            if (!isNullOrEmpty(kubernetesConfig.getNamespace())) {
+                kubernetesEndpointGroupBuilder.namespace(kubernetesConfig.getNamespace());
+            }
+            if (!isNullOrEmpty(watcher.getPortName())) {
+                kubernetesEndpointGroupBuilder.portName(watcher.getPortName());
+            }
+            // This callback can be executed by an event loop from CachingRepository, so we should use the
+            // specified executor to avoid blocking the event loop below.
+            return kubernetesEndpointGroupBuilder.build();
+        }, executor);
     }
 
-    private static Config toConfig(KubernetesConfig kubernetesConfig) {
+    private static CompletableFuture<Config> toConfig(KubernetesConfig kubernetesConfig,
+                                                      MetaRepository metaRepository) {
         final ConfigBuilder configBuilder = new ConfigBuilder()
                 .withMasterUrl(kubernetesConfig.getControlPlaneUrl())
                 .withTrustCerts(kubernetesConfig.getTrustCerts());
 
-        if (!isNullOrEmpty(kubernetesConfig.getOauthToken())) {
-            configBuilder.withOauthToken(kubernetesConfig.getOauthToken());
+        final String oauthToken = kubernetesConfig.getOauthToken();
+        if (isNullOrEmpty(oauthToken)) {
+            return CompletableFuture.completedFuture(configBuilder.build());
         }
 
-        return configBuilder.build();
+        if (!oauthToken.startsWith("credential:")) {
```

Comment:
Question) Would it make more sense to enforce users to use credentials?

Since we're still free to make breaking changes, we could just rename `oauthToken` to `credentialId` or `oauthCredentialId` while we're at it.

Reply:
Yeah, that's a good idea. 👍

Reply:
Addressed. 😉